Fan: preset_mode not working anymore

### The problem

The Fan component has breaking changes in 2025.11 without announcing it here:
https://esphome.io/changelog/2025.11.0/

preset_mode was publicly available and used in examples, here:
```
if(id(my_fan).preset_mode == "auto")  {}
```

This seems to be the change in question:
https://github.com/esphome/esphome/commit/885508775f6c1a556410c462b95f247c6b249d05


### Which version of ESPHome has the issue?

2025.11.0+

### What type of installation are you using?

pip

### What platform are you using?

ESP32

### Component causing the issue

fan

### YAML Config

```yaml
fan:
  - platform: template
    name: "Fan"
    id: fan_lv131
    speed_count: 3
    preset_modes:
      - Manual
      - Auto
      - Sleep    
    on_state:
    - logger.log:
        format: "Fan State changed!  Fan Speed is %d!"
        args: [ x->speed ]
    - logger.log:
        format: "Fan State changed!  Fan Mode is %s!"
        args: [ x->preset_mode.c_str()]
    - logger.log:
        format: "Fan State changed!  Fan state is %d!"
        args: [ x->state ]
```

### Anything in the logs that might be useful for us?

```txt
This is the error i get after using 2025.11+
In file included from src/esphome/core/component.h:8,
                 from src/esphome/core/application.h:7,
                 from src/esphome\components\api\api_frame_helper.h:13,
                 from src/esphome\components\api\api_connection.h:5,
                 from src/esphome.h:3,
                 from src/main.cpp:3:
src/main.cpp: In lambda function:
src/main.cpp:532:66: error: 'class esphome::fan::Fan' has no member named 'preset_mode'; did you mean 'const char* esphome::fan::Fan::preset_mode_'? (accessible via 'const char* esphome::fan::Fan::get_preset_mode() const')
  532 |       ESP_LOGD("main", "Fan State changed!  Fan Mode is %s!", x->preset_mode.c_str());
      |                                                                  ^~~~~~~~~~~
src/esphome/core/log.h:99:100: note: in definition of macro 'esph_log_d'
   99 |   ::esphome::esp_log_printf_(ESPHOME_LOG_LEVEL_DEBUG, tag, __LINE__, ESPHOME_LOG_FORMAT(format), ##__VA_ARGS__)
      |                                                                                                    ^~~~~~~~~~~
src/main.cpp:532:7: note: in expansion of macro 'ESP_LOGD'
  532 |       ESP_LOGD("main", "Fan State changed!  Fan Mode is %s!", x->preset_mode.c_str());
      |       ^~~~~~~~
levoit-lv131s.yaml: In lambda function:
levoit-lv131s.yaml:387:24: error: 'class esphome::template_::TemplateFan' has no member named 'preset_mode'; did you mean 'const char* esphome::fan::Fan::preset_mode_'? (accessible via 'const char* esphome::fan::Fan::get_preset_mode() const')
levoit-lv131s.yaml:395:24: error: 'class esphome::template_::TemplateFan' has no member named 'preset_mode'; did you mean 'const char* esphome::fan::Fan::preset_mode_'? (accessible via 'const char* esphome::fan::Fan::get_preset_mode() const')
```

### Additional information

When I replace 

`id(fan_lv131)->preset_mode `
with
`id(fan_lv131)->get_preset_mode() `
it craches!

What works is this
`['id(fan_lv131)->get_preset_mode() != nullptr ? x->get_preset_mode() : ""']`

### Suggested fix

i would suggest adding the nullptr handling to get_preset_mode so that it behaves as .preset_mode (return empty string instead off nullptr)

in fan.h
```
  /// Get the current preset mode (returns pointer to string stored in traits, or empty char if not set)
  const char *get_preset_mode() const { return this->preset_mode_ != nullptr ?this->preset_mode_:""; }
```
I verified that after applying this fix to 2025.12.4 i can simply replace .preset_mode with .get_preset_mode();

The docs need to be updated if this breaking changes stays (unless it is possible to provide .preset_mode like before without more ram)


### Logs for Guru Meditation 

```
[11:52:49.136][Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.

[11:52:49.137]Core  1 register dump:
[11:52:49.139]PC      : 0x400014fd  PS      : 0x00060e30  A0      : 0x80154e55  A1      : 0x3ffbfa70
[11:52:49.159]A2      : 0x00000000  A3      : 0xfffffffc  A4      : 0x000000ff  A5      : 0x0000ff00  
[11:52:49.160]A6      : 0x00ff0000  A7      : 0xff000000  A8      : 0x00000000  A9      : 0x3ffbffd0
[11:52:49.162]A10     : 0x00000005  A11     : 0x3f402ada  A12     : 0x00000214  A13     : 0x3f402adf
[11:52:49.163]A14     : 0x00000000  A15     : 0x00000000  SAR     : 0x00000010  EXCCAUSE: 0x0000001c
[11:52:49.165]EXCVADDR: 0x00000000  LBEG    : 0x400014fd  LEND    : 0x4000150d  LCOUNT  : 0xffffffff


[11:52:49.173]Backtrace: 0x400014fa:0x3ffbfa70 0x40154e52:0x3ffbfa80 0x4015098d:0x3ffbfda0 0x401509ca:0x3ffbfe30 0x400d8fbd:0x3ffbfe70 0x400d91d6:0x3ffbfec0 0x400e02b6:0x3ffbffa0 0x400e02e2:0x3ffbffd0 0x400e131a:0x3ffc0030 0x4015e227:0x3ffc0050 0x4015e171:0x3ffc0070 0x4015e187:0x3ffc0090 0x4015e4c3:0x3ffc00b0 0x400d85f2:0x3ffc00d0 0x400d8676:0x3ffc0110 0x400dbcc9:0x3ffc0150 0x4015dedd:0x3ffc01b0 0x400df5c8:0x3ffc01d0 0x400deefb:0x3ffc01f0 0x400e3056:0x3ffc0230 0x400d6ceb:0x3ffc02f0
WARNING Found stack trace! Trying to decode it
WARNING Decoded 0x40154e52: _svfprintf_r at /builds/idf/crosstool-NG/.build/xtensa-esp-elf/src/newlib/newlib/libc/stdio/vfprintf.c:1521
WARNING Decoded 0x4015098d: _vsnprintf_r at /builds/idf/crosstool-NG/.build/xtensa-esp-elf/src/newlib/newlib/libc/stdio/vsnprintf.c:70
WARNING Decoded 0x401509ca: vsnprintf at /builds/idf/crosstool-NG/.build/xtensa-esp-elf/src/newlib/newlib/libc/stdio/vsnprintf.c:40 (discriminator 1)
WARNING Decoded 0x400d8fbd: esphome::logger::Logger::format_body_to_buffer_(char*, unsigned short*, unsigned short, char const*, __va_list_tag) at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/esphome/components/logger/logger.h:421
 (inlined by) esphome::logger::Logger::format_log_to_buffer_with_terminator_(unsigned char, char const*, int, char const*, __va_list_tag, char*, unsigned short*, unsigned short) at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/esphome/components/logger/logger.h:191
WARNING Decoded 0x400d91d6: esphome::logger::Logger::log_message_to_buffer_and_send_(unsigned char, char const*, int, char const*, __va_list_tag) at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/esphome/components/logger/logger.h:208
 (inlined by) esphome::logger::Logger::log_vprintf_(unsigned char, char const*, int, char const*, __va_list_tag) at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/esphome/components/logger/logger.cpp:40
WARNING Decoded 0x400e02b6: esphome::esp_log_vprintf_(int, char const*, int, char const*, __va_list_tag) at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/esphome/core/log.cpp:32
WARNING Decoded 0x400e02e2: esphome::esp_log_printf_(int, char const*, int, char const*, ...) at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/esphome/core/log.cpp:14
WARNING Decoded 0x400e131a: setup()::{lambda(esphome::fan::Fan*)#2}::_FUN(esphome::fan::Fan*) at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/main.cpp:532 (discriminator 1)
 (inlined by) _FUN at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/main.cpp:533 (discriminator 1)
WARNING Decoded 0x4015e227: esphome::StatelessLambdaAction<esphome::fan::Fan*>::play(esphome::fan::Fan* const&) at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/esphome/core/base_automation.h:219
WARNING Decoded 0x4015e171: esphome::Action<esphome::fan::Fan*>::play_complex(esphome::fan::Fan* const&) at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/esphome/core/automation.h:199
WARNING Decoded 0x4015e187: esphome::Action<esphome::fan::Fan*>::play_next_(esphome::fan::Fan* const&) at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/esphome/core/automation.h:230
 (inlined by) esphome::Action<esphome::fan::Fan*>::play_complex(esphome::fan::Fan* const&) at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/esphome/core/automation.h:200
WARNING Decoded 0x4015e4c3: esphome::ActionList<esphome::fan::Fan*>::play(esphome::fan::Fan* const&) at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/esphome/core/automation.h:278
 (inlined by) esphome::Automation<esphome::fan::Fan*>::trigger(esphome::fan::Fan* const&) at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/esphome/core/automation.h:318
 (inlined by) esphome::Trigger<esphome::fan::Fan*>::trigger(esphome::fan::Fan* const&) at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/esphome/core/automation.h:172
 (inlined by) esphome::fan::FanStateTrigger::FanStateTrigger(esphome::fan::Fan*)::{lambda()#1}::operator()() const at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/esphome\components\fan/automation.h:117
 (inlined by) void std::__invoke_impl<void, esphome::fan::FanStateTrigger::FanStateTrigger(esphome::fan::Fan*)::{lambda()#1}&>(std::__invoke_other, esphome::fan::FanStateTrigger::FanStateTrigger(esphome::fan::Fan*)::{lambda()#1}&) at C:/Users/tobis/.platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/invoke.h:61   
 (inlined by) std::enable_if<is_invocable_r_v<void, esphome::fan::FanStateTrigger::FanStateTrigger(esphome::fan::Fan*)::{lambda()#1}&>, void>::type std::__invoke_r<void, esphome::fan::FanStateTrigger::FanStateTrigger(esphome::fan::Fan*)::{lambda()#1}&>(esphome::fan::FanStateTrigger::FanStateTrigger(esphome::fan::Fan*)::{lambda()#1}&) at C:/Users/tobis/.platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/invoke.h:111
 (inlined by) std::_Function_handler<void (), esphome::fan::FanStateTrigger::FanStateTrigger(esphome::fan::Fan*)::{lambda()#1}>::_M_invoke(std::_Any_data const&) at C:/Users/tobis/.platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/std_function.h:290
WARNING Decoded 0x400d85f2: std::function<void ()>::operator()() const at C:/Users/tobis/.platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/std_function.h:591
 (inlined by) esphome::CallbackManager<void ()>::call() at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/esphome/core/helpers.h:864
 (inlined by) esphome::fan::Fan::publish_state() at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/esphome/components/fan/fan.cpp:185
WARNING Decoded 0x400d8676: esphome::fan::FanRestoreState::apply(esphome::fan::Fan&) at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/esphome/components/fan/fan.cpp:135
WARNING Decoded 0x400dbcc9: esphome::template_::TemplateFan::setup() at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/esphome/components/template/fan/template_fan.cpp:12 (discriminator 1)
WARNING Decoded 0x4015dedd: esphome::Component::call_setup() at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/esphome/core/component.cpp:140
WARNING Decoded 0x400df5c8: esphome::Component::call() at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/esphome/core/component.cpp:170
WARNING Decoded 0x400deefb: esphome::Application::setup() at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/esphome/core/application.cpp:88
WARNING Decoded 0x400e3056: setup() at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/levoit-lv131s.yaml:21
WARNING Decoded 0x400d6ceb: esphome::loop_task(void*) at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/esphome/components/esp32/core.cpp:92




[11:52:52.312]ELF file SHA256: 86e49c322

[11:52:52.313]Rebooting...
[11:52:52.313]ets Jul 29 2019 12:21:46

[11:52:52.314]rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
[11:52:52.315]configsip: 0, SPIWP:0xee
[11:52:52.316]clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
[11:52:52.317]mode:DIO, clock div:2
[11:52:52.317]load:0x3fff0030,len:6380
[11:52:52.318]ho 0 tail 12 room 4
[11:52:52.318]load:0x40078000,len:15916
[11:52:52.319]load:0x40080400,len:3860
[11:52:52.320]entry 0x40080630
[11:52:52.321]I (29) boot: ESP-IDF 5.5.1 2nd stage bootloader
[11:52:52.321]I (29) boot: compile time Jan  1 2026 11:50:31
[11:52:52.322]I (29) boot: Multicore bootloader
[11:52:52.323]I (30) boot: chip revision: v3.1
[11:52:52.324]I (33) boot.esp32: SPI Speed      : 40MHz
[11:52:52.325]I (37) boot.esp32: SPI Mode       : DIO
[11:52:52.325]I (40) boot.esp32: SPI Flash Size : 4MB
[11:52:52.326]I (44) boot: Enabling RNG early entropy source...
[11:52:52.327]I (49) boot: Partition Table:
[11:52:52.328]I (51) boot: ## Label            Usage          Type ST Offset   Length
[11:52:52.330]I (57) boot:  0 otadata          OTA data         01 00 00009000 00002000
[11:52:52.331]I (64) boot:  1 phy_init         RF data          01 01 0000b000 00001000
[11:52:52.333]I (71) boot:  2 app0             OTA app          00 10 00010000 001c0000
[11:52:52.334]I (77) boot:  3 app1             OTA app          00 11 001d0000 001c0000
[11:52:52.335]I (84) boot:  4 nvs              WiFi data        01 02 00390000 0006d000
[11:52:52.336]I (90) boot: End of partition table
[11:52:52.338]I (93) esp_image: segment 0: paddr=00010020 vaddr=3f400020 size=23a9ch (146076) map
[11:52:52.339]I (151) esp_image: segment 1: paddr=00033ac4 vaddr=3ffb0000 size=03fc4h ( 16324) load
[11:52:52.341]I (157) esp_image: segment 2: paddr=00037a90 vaddr=40080000 size=08588h ( 34184) load
[11:52:52.342]I (171) esp_image: segment 3: paddr=00040020 vaddr=400d0020 size=96ffch (618492) map
[11:52:52.344]I (382) esp_image: segment 4: paddr=000d7024 vaddr=40088588 size=103a4h ( 66468) load
[11:52:52.346]I (408) esp_image: segment 5: paddr=000e73d0 vaddr=50000000 size=00020h (    32) load
[11:52:52.347]I (420) boot: Loaded app from partition at offset 0x10000
[11:52:52.348]I (421) boot: Disabling RNG early entropy source...
[11:52:52.349][I][logger:121]: Log initialized
[11:52:52.350][C][safe_mode:084]: Unsuccessful boot attempts: 2
[11:52:52.351][D][esp32.preferences:149]: Writing 1 items: 0 cached, 1 written, 0 failed
[11:52:52.352][I][app:073]: Running through setup()
[11:52:52.353][C][component:173]: Setup spi took 0ms
[11:52:52.355][C][component:173]: Setup uart took 1ms
[11:52:52.356][C][component:173]: Setup preferences took 1ms
[11:52:52.357][C][component:173]: Setup ledc.output took 1ms
[11:52:52.358][C][component:173]: Setup gpio.output took 0ms
[11:52:52.359][C][component:173]: Setup ledc.output took 0ms
[11:52:52.360][C][component:173]: Setup template.sensor took 0ms
[11:52:52.362][C][component:173]: Setup template.binary_sensor took 0ms
[11:52:52.363][C][component:173]: Setup template.binary_sensor took 0ms
[11:52:52.364][C][component:173]: Setup template.binary_sensor took 0ms
[11:52:52.365][C][component:173]: Setup template.binary_sensor took 0ms
[11:52:52.367][C][component:173]: Setup template.binary_sensor took 0ms
[11:52:52.368][C][component:173]: Setup script took 0ms
[11:52:52.369][C][component:173]: Setup template.binary_sensor took 0ms
[11:52:52.370][D][template.switch:045]:   Restored state ON
[11:52:52.371][D][switch:022]: 'Display' Turning ON.
[11:52:52.372][D][switch:065]: 'Display': Sending state ON
[11:52:52.373][C][component:173]: Setup template.switch took 20ms
[11:52:52.374][D][template.switch:045]:   Restored state ON
[11:52:52.375][D][switch:022]: 'Sound' Turning ON.
[11:52:52.376][D][switch:065]: 'Sound': Sending state ON
[11:52:52.377][C][component:173]: Setup template.switch took 20ms
[11:52:52.379][D][spi_device:393]: mode 0, data_rate 200kHz
[11:52:52.380][C][component:173]: Setup spi_device took 11ms
[11:52:52.381][C][component:173]: Setup uart took 0ms
[11:52:52.382][Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.

[11:52:52.383]Core  1 register dump:
[11:52:52.385]PC      : 0x400014fd  PS      : 0x00060e30  A0      : 0x80154e55  A1      : 0x3ffbfa70
[11:52:52.405]A2      : 0x00000000  A3      : 0xfffffffc  A4      : 0x000000ff  A5      : 0x0000ff00  
[11:52:52.407]A6      : 0x00ff0000  A7      : 0xff000000  A8      : 0x00000000  A9      : 0x3ffbffd0  
[11:52:52.409]A10     : 0x00000005  A11     : 0x3f402ada  A12     : 0x00000214  A13     : 0x3f402adf
[11:52:52.410]A14     : 0x00000000  A15     : 0x00000000  SAR     : 0x00000010  EXCCAUSE: 0x0000001c
[11:52:52.412]EXCVADDR: 0x00000000  LBEG    : 0x400014fd  LEND    : 0x4000150d  LCOUNT  : 0xffffffff


[11:52:52.420]Backtrace: 0x400014fa:0x3ffbfa70 0x40154e52:0x3ffbfa80 0x4015098d:0x3ffbfda0 0x401509ca:0x3ffbfe30 0x400d8fbd:0x3ffbfe70 0x400d91d6:0x3ffbfec0 0x400e02b6:0x3ffbffa0 0x400e02e2:0x3ffbffd0 0x400e131a:0x3ffc0030 0x4015e227:0x3ffc0050 0x4015e171:0x3ffc0070 0x4015e187:0x3ffc0090 0x4015e4c3:0x3ffc00b0 0x400d85f2:0x3ffc00d0 0x400d8676:0x3ffc0110 0x400dbcc9:0x3ffc0150 0x4015dedd:0x3ffc01b0 0x400df5c8:0x3ffc01d0 0x400deefb:0x3ffc01f0 0x400e3056:0x3ffc0230 0x400d6ceb:0x3ffc02f0
WARNING Found stack trace! Trying to decode it
WARNING Decoded 0x40154e52: _svfprintf_r at /builds/idf/crosstool-NG/.build/xtensa-esp-elf/src/newlib/newlib/libc/stdio/vfprintf.c:1521
WARNING Decoded 0x4015098d: _vsnprintf_r at /builds/idf/crosstool-NG/.build/xtensa-esp-elf/src/newlib/newlib/libc/stdio/vsnprintf.c:70
WARNING Decoded 0x401509ca: vsnprintf at /builds/idf/crosstool-NG/.build/xtensa-esp-elf/src/newlib/newlib/libc/stdio/vsnprintf.c:40 (discriminator 1)
WARNING Decoded 0x400d8fbd: esphome::logger::Logger::format_body_to_buffer_(char*, unsigned short*, unsigned short, char const*, __va_list_tag) at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/esphome/components/logger/logger.h:421
 (inlined by) esphome::logger::Logger::format_log_to_buffer_with_terminator_(unsigned char, char const*, int, char const*, __va_list_tag, char*, unsigned short*, unsigned short) at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/esphome/components/logger/logger.h:191
WARNING Decoded 0x400d91d6: esphome::logger::Logger::log_message_to_buffer_and_send_(unsigned char, char const*, int, char const*, __va_list_tag) at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/esphome/components/logger/logger.h:208
 (inlined by) esphome::logger::Logger::log_vprintf_(unsigned char, char const*, int, char const*, __va_list_tag) at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/esphome/components/logger/logger.cpp:40
WARNING Decoded 0x400e02b6: esphome::esp_log_vprintf_(int, char const*, int, char const*, __va_list_tag) at C:\Users\tobis\Documents\esphome\esphome-projects\projects\levoit-lv131s\.esphome\build\levoit-lv131s/src/esphome/core/log.cpp:32
```

--- a/esphome/components/api/api_connection.cpp
+++ b/esphome/components/api/api_connection.cpp
@@ -442,8 +442,10 @@ uint16_t APIConnection::try_send_fan_state(EntityBase *entity, APIConnection *co
   }
   if (traits.supports_direction())
     msg.direction = static_cast<enums::FanDirection>(fan->direction);
-  if (traits.supports_preset_modes() && fan->has_preset_mode())
-    msg.preset_mode = StringRef(fan->get_preset_mode());
+  if (traits.supports_preset_modes() && fan->has_preset_mode()) {
+    auto preset = fan->get_preset_mode();
+    msg.preset_mode = StringRef(preset.data(), preset.size());
+  }
   return fill_and_encode_entity_state(fan, msg, FanStateResponse::MESSAGE_TYPE, conn, remaining_size, is_single);
 }
 uint16_t APIConnection::try_send_fan_info(EntityBase *entity, APIConnection *conn, uint32_t remaining_size,

--- a/esphome/components/api/api_connection.cpp
+++ b/esphome/components/api/api_connection.cpp
@@ -442,10 +442,8 @@ uint16_t APIConnection::try_send_fan_state(EntityBase *entity, APIConnection *co
   }
   if (traits.supports_direction())
     msg.direction = static_cast<enums::FanDirection>(fan->direction);
-  if (traits.supports_preset_modes() && fan->has_preset_mode()) {
-    auto preset = fan->get_preset_mode();
-    msg.preset_mode = StringRef(preset.data(), preset.size());
-  }
+  if (traits.supports_preset_modes() && fan->has_preset_mode())
+    msg.preset_mode = fan->get_preset_mode();
   return fill_and_encode_entity_state(fan, msg, FanStateResponse::MESSAGE_TYPE, conn, remaining_size, is_single);
 }
 uint16_t APIConnection::try_send_fan_info(EntityBase *entity, APIConnection *conn, uint32_t remaining_size,

--- a/esphome/components/copy/fan/copy_fan.cpp
+++ b/esphome/components/copy/fan/copy_fan.cpp
@@ -8,20 +8,24 @@ static const char *const TAG = "copy.fan";
 
 void CopyFan::setup() {
   source_->add_on_state_callback([this]() {
-    this->state = source_->state;
-    this->oscillating = source_->oscillating;
-    this->speed = source_->speed;
-    this->direction = source_->direction;
-    this->set_preset_mode_(source_->get_preset_mode());
+    this->copy_state_from_source_();
     this->publish_state();
   });
 
+  this->copy_state_from_source_();
+  this->publish_state();
+}
+
+void CopyFan::copy_state_from_source_() {
   this->state = source_->state;
   this->oscillating = source_->oscillating;
   this->speed = source_->speed;
   this->direction = source_->direction;
-  this->set_preset_mode_(source_->get_preset_mode());
-  this->publish_state();
+  if (source_->has_preset_mode()) {
+    this->set_preset_mode_(source_->get_preset_mode());
+  } else {
+    this->clear_preset_mode_();
+  }
 }
 
 void CopyFan::dump_config() { LOG_FAN("", "Copy Fan", this); }

--- a/esphome/components/copy/fan/copy_fan.h
+++ b/esphome/components/copy/fan/copy_fan.h
@@ -16,7 +16,7 @@ class CopyFan : public fan::Fan, public Component {
 
  protected:
   void control(const fan::FanCall &call) override;
-  ;
+  void copy_state_from_source_();
 
   fan::Fan *source_;
 };

--- a/esphome/components/fan/automation.h
+++ b/esphome/components/fan/automation.h
@@ -212,18 +212,18 @@ class FanPresetSetTrigger : public Trigger<std::string> {
  public:
   FanPresetSetTrigger(Fan *state) {
     state->add_on_state_callback([this, state]() {
-      const auto *preset_mode = state->get_preset_mode();
+      auto preset_mode = state->get_preset_mode();
       auto should_trigger = preset_mode != this->last_preset_mode_;
       this->last_preset_mode_ = preset_mode;
       if (should_trigger) {
-        this->trigger(preset_mode);
+        this->trigger(std::string(preset_mode));
       }
     });
     this->last_preset_mode_ = state->get_preset_mode();
   }
 
  protected:
-  const char *last_preset_mode_{""};
+  std::string_view last_preset_mode_{};
 };
 
 }  // namespace fan

--- a/esphome/components/fan/automation.h
+++ b/esphome/components/fan/automation.h
@@ -223,7 +223,7 @@ class FanPresetSetTrigger : public Trigger<std::string> {
   }
 
  protected:
-  std::string_view last_preset_mode_{};
+  StringRef last_preset_mode_{};
 };
 
 }  // namespace fan

--- a/esphome/components/fan/automation.h
+++ b/esphome/components/fan/automation.h
@@ -216,15 +216,14 @@ class FanPresetSetTrigger : public Trigger<std::string> {
       auto should_trigger = preset_mode != this->last_preset_mode_;
       this->last_preset_mode_ = preset_mode;
       if (should_trigger) {
-        // Trigger with empty string when nullptr to maintain backward compatibility
-        this->trigger(preset_mode != nullptr ? preset_mode : "");
+        this->trigger(preset_mode);
       }
     });
     this->last_preset_mode_ = state->get_preset_mode();
   }
 
  protected:
-  const char *last_preset_mode_{nullptr};
+  const char *last_preset_mode_{""};
 };
 
 }  // namespace fan

--- a/esphome/components/fan/fan.cpp
+++ b/esphome/components/fan/fan.cpp
@@ -154,16 +154,16 @@ const char *Fan::find_preset_mode_(const char *preset_mode, size_t len) {
   return this->get_traits().find_preset_mode(preset_mode, len);
 }
 
-bool Fan::set_preset_mode_(const char *preset_mode) {
-  if (preset_mode == nullptr) {
-    // Treat nullptr as clearing the preset mode
+bool Fan::set_preset_mode_(const char *preset_mode, size_t len) {
+  if (preset_mode == nullptr || len == 0) {
+    // Treat nullptr/empty as clearing the preset mode
     if (this->preset_mode_ == nullptr) {
       return false;  // No change
     }
     this->clear_preset_mode_();
     return true;
   }
-  const char *validated = this->find_preset_mode_(preset_mode);
+  const char *validated = this->find_preset_mode_(preset_mode, len);
   if (validated == nullptr || this->preset_mode_ == validated) {
     return false;  // Preset mode not supported or no change
   }
@@ -171,7 +171,17 @@ bool Fan::set_preset_mode_(const char *preset_mode) {
   return true;
 }
 
-bool Fan::set_preset_mode_(const std::string &preset_mode) { return this->set_preset_mode_(preset_mode.c_str()); }
+bool Fan::set_preset_mode_(const char *preset_mode) {
+  return this->set_preset_mode_(preset_mode, preset_mode ? strlen(preset_mode) : 0);
+}
+
+bool Fan::set_preset_mode_(const std::string &preset_mode) {
+  return this->set_preset_mode_(preset_mode.data(), preset_mode.size());
+}
+
+bool Fan::set_preset_mode_(std::string_view preset_mode) {
+  return this->set_preset_mode_(preset_mode.data(), preset_mode.size());
+}
 
 void Fan::clear_preset_mode_() { this->preset_mode_ = nullptr; }
 

--- a/esphome/components/fan/fan.cpp
+++ b/esphome/components/fan/fan.cpp
@@ -61,7 +61,7 @@ void FanCall::perform() {
   if (this->direction_.has_value()) {
     ESP_LOGD(TAG, "  Direction: %s", LOG_STR_ARG(fan_direction_to_string(*this->direction_)));
   }
-  if (this->has_preset_mode()) {
+  if (this->preset_mode_ != nullptr) {
     ESP_LOGD(TAG, "  Preset Mode: %s", this->preset_mode_);
   }
   this->parent_.control(*this);
@@ -201,8 +201,8 @@ void Fan::publish_state() {
   if (traits.supports_direction()) {
     ESP_LOGD(TAG, "  Direction: %s", LOG_STR_ARG(fan_direction_to_string(this->direction)));
   }
-  if (this->has_preset_mode()) {
-    ESP_LOGD(TAG, "  Preset Mode: %s", this->get_preset_mode());
+  if (this->preset_mode_ != nullptr) {
+    ESP_LOGD(TAG, "  Preset Mode: %s", this->preset_mode_);
   }
   this->state_callback_.call();
 #if defined(USE_FAN) && defined(USE_CONTROLLER_REGISTRY)

--- a/esphome/components/fan/fan.cpp
+++ b/esphome/components/fan/fan.cpp
@@ -83,7 +83,7 @@ void FanCall::validate_() {
       *this->binary_state_
       // ..,and no preset mode will be active...
       && !this->has_preset_mode() &&
-      this->parent_.get_preset_mode() == nullptr
+      !this->parent_.has_preset_mode()
       // ...and neither current nor new speed is available...
       && traits.supports_speed() && this->parent_.speed == 0 && !this->speed_.has_value()) {
     // ...set speed to 100%
@@ -175,6 +175,15 @@ bool Fan::set_preset_mode_(const std::string &preset_mode) { return this->set_pr
 
 void Fan::clear_preset_mode_() { this->preset_mode_ = nullptr; }
 
+void Fan::apply_preset_mode_(const FanCall &call) {
+  if (call.has_preset_mode()) {
+    this->set_preset_mode_(call.get_preset_mode());
+  } else if (call.get_speed().has_value()) {
+    // Manually setting speed clears preset (per Home Assistant convention)
+    this->clear_preset_mode_();
+  }
+}
+
 void Fan::add_on_state_callback(std::function<void()> &&callback) { this->state_callback_.add(std::move(callback)); }
 void Fan::publish_state() {
   auto traits = this->get_traits();
@@ -192,9 +201,8 @@ void Fan::publish_state() {
   if (traits.supports_direction()) {
     ESP_LOGD(TAG, "  Direction: %s", LOG_STR_ARG(fan_direction_to_string(this->direction)));
   }
-  const char *preset = this->get_preset_mode();
-  if (preset != nullptr) {
-    ESP_LOGD(TAG, "  Preset Mode: %s", preset);
+  if (this->has_preset_mode()) {
+    ESP_LOGD(TAG, "  Preset Mode: %s", this->get_preset_mode());
   }
   this->state_callback_.call();
 #if defined(USE_FAN) && defined(USE_CONTROLLER_REGISTRY)
@@ -249,12 +257,11 @@ void Fan::save_state_() {
   state.speed = this->speed;
   state.direction = this->direction;
 
-  const char *preset = this->get_preset_mode();
-  if (preset != nullptr) {
+  if (this->has_preset_mode()) {
     const auto &preset_modes = traits.supported_preset_modes();
     // Find index of current preset mode (pointer comparison is safe since preset is from traits)
     for (size_t i = 0; i < preset_modes.size(); i++) {
-      if (preset_modes[i] == preset) {
+      if (preset_modes[i] == this->preset_mode_) {
         state.preset_mode = i;
         break;
       }

--- a/esphome/components/fan/fan.cpp
+++ b/esphome/components/fan/fan.cpp
@@ -156,7 +156,7 @@ const char *Fan::find_preset_mode_(const char *preset_mode, size_t len) {
 
 bool Fan::set_preset_mode_(const char *preset_mode, size_t len) {
   if (preset_mode == nullptr || len == 0) {
-    // Treat nullptr/empty as clearing the preset mode
+    // Treat nullptr or empty string as clearing the preset mode (no valid preset is "")
     if (this->preset_mode_ == nullptr) {
       return false;  // No change
     }
@@ -180,6 +180,8 @@ bool Fan::set_preset_mode_(const std::string &preset_mode) {
 }
 
 bool Fan::set_preset_mode_(std::string_view preset_mode) {
+  // Safe: find_preset_mode_ only uses the input for comparison and returns
+  // a pointer from traits, so the input string_view's lifetime doesn't matter.
   return this->set_preset_mode_(preset_mode.data(), preset_mode.size());
 }
 

--- a/esphome/components/fan/fan.cpp
+++ b/esphome/components/fan/fan.cpp
@@ -179,10 +179,10 @@ bool Fan::set_preset_mode_(const std::string &preset_mode) {
   return this->set_preset_mode_(preset_mode.data(), preset_mode.size());
 }
 
-bool Fan::set_preset_mode_(std::string_view preset_mode) {
+bool Fan::set_preset_mode_(StringRef preset_mode) {
   // Safe: find_preset_mode_ only uses the input for comparison and returns
-  // a pointer from traits, so the input string_view's lifetime doesn't matter.
-  return this->set_preset_mode_(preset_mode.data(), preset_mode.size());
+  // a pointer from traits, so the input StringRef's lifetime doesn't matter.
+  return this->set_preset_mode_(preset_mode.c_str(), preset_mode.size());
 }
 
 void Fan::clear_preset_mode_() { this->preset_mode_ = nullptr; }

--- a/esphome/components/fan/fan.h
+++ b/esphome/components/fan/fan.h
@@ -1,12 +1,11 @@
 #pragma once
 
-#include <string_view>
-
 #include "esphome/core/entity_base.h"
 #include "esphome/core/helpers.h"
 #include "esphome/core/log.h"
 #include "esphome/core/optional.h"
 #include "esphome/core/preferences.h"
+#include "esphome/core/string_ref.h"
 #include "fan_traits.h"
 
 namespace esphome {
@@ -131,12 +130,10 @@ class Fan : public EntityBase {
   void set_restore_mode(FanRestoreMode restore_mode) { this->restore_mode_ = restore_mode; }
 
   /// Get the current preset mode.
-  /// Returns a view of the string stored in traits, or empty view if not set.
-  /// The returned view points to string literals from codegen (static storage).
+  /// Returns a StringRef of the string stored in traits, or empty ref if not set.
+  /// The returned ref points to string literals from codegen (static storage).
   /// Traits are set once at startup and valid for the lifetime of the program.
-  std::string_view get_preset_mode() const {
-    return this->preset_mode_ != nullptr ? std::string_view(this->preset_mode_) : std::string_view();
-  }
+  StringRef get_preset_mode() const { return StringRef::from_maybe_nullptr(this->preset_mode_); }
 
   /// Check if a preset mode is currently active
   bool has_preset_mode() const { return this->preset_mode_ != nullptr; }
@@ -157,7 +154,7 @@ class Fan : public EntityBase {
   bool set_preset_mode_(const char *preset_mode, size_t len);
   bool set_preset_mode_(const char *preset_mode);
   bool set_preset_mode_(const std::string &preset_mode);
-  bool set_preset_mode_(std::string_view preset_mode);
+  bool set_preset_mode_(StringRef preset_mode);
   /// Clear the preset mode
   void clear_preset_mode_();
   /// Apply preset mode from a FanCall (handles speed-clears-preset convention)

--- a/esphome/components/fan/fan.h
+++ b/esphome/components/fan/fan.h
@@ -130,7 +130,9 @@ class Fan : public EntityBase {
   /// Set the restore mode of this fan.
   void set_restore_mode(FanRestoreMode restore_mode) { this->restore_mode_ = restore_mode; }
 
-  /// Get the current preset mode (returns view of string stored in traits, or empty view if not set)
+  /// Get the current preset mode.
+  /// Returns a view of the string stored in traits (static storage), or empty view if not set.
+  /// Safe to use as the underlying string has static lifetime.
   std::string_view get_preset_mode() const {
     return this->preset_mode_ != nullptr ? std::string_view(this->preset_mode_) : std::string_view();
   }

--- a/esphome/components/fan/fan.h
+++ b/esphome/components/fan/fan.h
@@ -1,5 +1,7 @@
 #pragma once
 
+#include <string_view>
+
 #include "esphome/core/entity_base.h"
 #include "esphome/core/helpers.h"
 #include "esphome/core/log.h"
@@ -128,8 +130,10 @@ class Fan : public EntityBase {
   /// Set the restore mode of this fan.
   void set_restore_mode(FanRestoreMode restore_mode) { this->restore_mode_ = restore_mode; }
 
-  /// Get the current preset mode (returns pointer to string stored in traits, or empty string if not set)
-  const char *get_preset_mode() const { return this->preset_mode_ != nullptr ? this->preset_mode_ : ""; }
+  /// Get the current preset mode (returns view of string stored in traits, or empty view if not set)
+  std::string_view get_preset_mode() const {
+    return this->preset_mode_ != nullptr ? std::string_view(this->preset_mode_) : std::string_view();
+  }
 
   /// Check if a preset mode is currently active
   bool has_preset_mode() const { return this->preset_mode_ != nullptr; }

--- a/esphome/components/fan/fan.h
+++ b/esphome/components/fan/fan.h
@@ -150,9 +150,10 @@ class Fan : public EntityBase {
   void dump_traits_(const char *tag, const char *prefix);
 
   /// Set the preset mode (finds and stores pointer from traits). Returns true if changed.
+  bool set_preset_mode_(const char *preset_mode, size_t len);
   bool set_preset_mode_(const char *preset_mode);
-  /// Set the preset mode (finds and stores pointer from traits). Returns true if changed.
   bool set_preset_mode_(const std::string &preset_mode);
+  bool set_preset_mode_(std::string_view preset_mode);
   /// Clear the preset mode
   void clear_preset_mode_();
   /// Apply preset mode from a FanCall (handles speed-clears-preset convention)

--- a/esphome/components/fan/fan.h
+++ b/esphome/components/fan/fan.h
@@ -128,8 +128,8 @@ class Fan : public EntityBase {
   /// Set the restore mode of this fan.
   void set_restore_mode(FanRestoreMode restore_mode) { this->restore_mode_ = restore_mode; }
 
-  /// Get the current preset mode (returns pointer to string stored in traits, or nullptr if not set)
-  const char *get_preset_mode() const { return this->preset_mode_; }
+  /// Get the current preset mode (returns pointer to string stored in traits, or empty string if not set)
+  const char *get_preset_mode() const { return this->preset_mode_ != nullptr ? this->preset_mode_ : ""; }
 
   /// Check if a preset mode is currently active
   bool has_preset_mode() const { return this->preset_mode_ != nullptr; }
@@ -151,6 +151,8 @@ class Fan : public EntityBase {
   bool set_preset_mode_(const std::string &preset_mode);
   /// Clear the preset mode
   void clear_preset_mode_();
+  /// Apply preset mode from a FanCall (handles speed-clears-preset convention)
+  void apply_preset_mode_(const FanCall &call);
   /// Find and return the matching preset mode pointer from traits, or nullptr if not found.
   const char *find_preset_mode_(const char *preset_mode);
   const char *find_preset_mode_(const char *preset_mode, size_t len);

--- a/esphome/components/fan/fan.h
+++ b/esphome/components/fan/fan.h
@@ -131,8 +131,9 @@ class Fan : public EntityBase {
   void set_restore_mode(FanRestoreMode restore_mode) { this->restore_mode_ = restore_mode; }
 
   /// Get the current preset mode.
-  /// Returns a view of the string stored in traits (static storage), or empty view if not set.
-  /// Safe to use as the underlying string has static lifetime.
+  /// Returns a view of the string stored in traits, or empty view if not set.
+  /// The returned view points to string literals from codegen (static storage).
+  /// Traits are set once at startup and valid for the lifetime of the program.
   std::string_view get_preset_mode() const {
     return this->preset_mode_ != nullptr ? std::string_view(this->preset_mode_) : std::string_view();
   }

--- a/esphome/components/fan/fan.h
+++ b/esphome/components/fan/fan.h
@@ -153,6 +153,7 @@ class Fan : public EntityBase {
   void dump_traits_(const char *tag, const char *prefix);
 
   /// Set the preset mode (finds and stores pointer from traits). Returns true if changed.
+  /// Passing nullptr or empty string clears the preset mode.
   bool set_preset_mode_(const char *preset_mode, size_t len);
   bool set_preset_mode_(const char *preset_mode);
   bool set_preset_mode_(const std::string &preset_mode);

--- a/esphome/components/hbridge/fan/hbridge_fan.cpp
+++ b/esphome/components/hbridge/fan/hbridge_fan.cpp
@@ -57,7 +57,7 @@ void HBridgeFan::control(const fan::FanCall &call) {
     this->oscillating = *call.get_oscillating();
   if (call.get_direction().has_value())
     this->direction = *call.get_direction();
-  this->set_preset_mode_(call.get_preset_mode());
+  this->apply_preset_mode_(call);
 
   this->write_state_();
   this->publish_state();

--- a/esphome/components/speed/fan/speed_fan.cpp
+++ b/esphome/components/speed/fan/speed_fan.cpp
@@ -29,7 +29,7 @@ void SpeedFan::control(const fan::FanCall &call) {
     this->oscillating = *call.get_oscillating();
   if (call.get_direction().has_value())
     this->direction = *call.get_direction();
-  this->set_preset_mode_(call.get_preset_mode());
+  this->apply_preset_mode_(call);
 
   this->write_state_();
   this->publish_state();

--- a/esphome/components/template/fan/template_fan.cpp
+++ b/esphome/components/template/fan/template_fan.cpp
@@ -28,7 +28,7 @@ void TemplateFan::control(const fan::FanCall &call) {
     this->oscillating = *call.get_oscillating();
   if (call.get_direction().has_value() && this->has_direction_)
     this->direction = *call.get_direction();
-  this->set_preset_mode_(call.get_preset_mode());
+  this->apply_preset_mode_(call);
 
   this->publish_state();
 }

--- a/tests/components/copy/common.yaml
+++ b/tests/components/copy/common.yaml
@@ -7,6 +7,9 @@ fan:
   - platform: speed
     id: fan_speed
     output: fan_output_1
+    preset_modes:
+      - Eco
+      - Turbo
   - platform: copy
     source_id: fan_speed
     name: Fan Speed Copy

--- a/tests/components/fan/common.yaml
+++ b/tests/components/fan/common.yaml
@@ -9,3 +9,27 @@ fan:
     has_oscillating: true
     has_direction: true
     speed_count: 3
+
+# Test lambdas using get_preset_mode() which returns std::string_view
+binary_sensor:
+  - platform: template
+    id: fan_has_preset
+    name: "Fan Has Preset"
+    lambda: |-
+      // Test has_preset_mode() method
+      if (!id(test_fan).has_preset_mode()) {
+        return false;
+      }
+      // Test .empty() on string_view
+      if (id(test_fan).get_preset_mode().empty()) {
+        return false;
+      }
+      // Test == comparison with string literal
+      if (id(test_fan).get_preset_mode() == "Eco") {
+        return true;
+      }
+      // Test != comparison
+      if (id(test_fan).get_preset_mode() != "Sleep") {
+        return true;
+      }
+      return false;

--- a/tests/components/fan/common.yaml
+++ b/tests/components/fan/common.yaml
@@ -11,23 +11,47 @@ fan:
     speed_count: 3
 
 # Test lambdas using get_preset_mode() which returns std::string_view
+# These examples match the migration guide in the PR description
 binary_sensor:
   - platform: template
     id: fan_has_preset
     name: "Fan Has Preset"
     lambda: |-
-      // Test has_preset_mode() method
-      if (!id(test_fan).has_preset_mode()) {
-        return false;
+      // Migration guide: Checking if preset mode is set
+      // Use empty() or has_preset_mode()
+      if (!id(test_fan).get_preset_mode().empty()) {
+        // preset is set
       }
-      // Test .empty() on string_view
-      if (id(test_fan).get_preset_mode().empty()) {
-        return false;
+      if (id(test_fan).has_preset_mode()) {
+        // preset is set
       }
-      // Test == comparison with string literal
+
+      // Migration guide: Comparing preset mode
+      // Use == operator directly (safe, works even when empty)
       if (id(test_fan).get_preset_mode() == "Eco") {
         return true;
       }
+
+      // Migration guide: Checking for no preset
+      if (id(test_fan).get_preset_mode().empty()) {
+        // no preset
+      }
+      if (!id(test_fan).has_preset_mode()) {
+        // no preset
+      }
+
+      // Migration guide: Getting as std::string
+      std::string preset = std::string(id(test_fan).get_preset_mode());
+
+      // Migration guide: Logging option 1
+      // Use .data() - works because string_view points to null-terminated string in traits
+      ESP_LOGD("test", "Preset: %s", id(test_fan).get_preset_mode().data());
+
+      // Migration guide: Logging option 2
+      // Use %.*s format (safer, no null-termination assumption)
+      auto preset_view = id(test_fan).get_preset_mode();
+      ESP_LOGD("test", "Preset: %.*s", (int)preset_view.size(), preset_view.data());
+
       // Test != comparison
       if (id(test_fan).get_preset_mode() != "Sleep") {
         return true;

--- a/tests/components/fan/common.yaml
+++ b/tests/components/fan/common.yaml
@@ -10,7 +10,7 @@ fan:
     has_direction: true
     speed_count: 3
 
-# Test lambdas using get_preset_mode() which returns std::string_view
+# Test lambdas using get_preset_mode() which returns StringRef
 # These examples match the migration guide in the PR description
 binary_sensor:
   - platform: template
@@ -44,13 +44,13 @@ binary_sensor:
       std::string preset = std::string(id(test_fan).get_preset_mode());
 
       // Migration guide: Logging option 1
-      // Use .data() - works because string_view points to null-terminated string in traits
-      ESP_LOGD("test", "Preset: %s", id(test_fan).get_preset_mode().data());
+      // Use .c_str() - works because StringRef points to null-terminated string in traits
+      ESP_LOGD("test", "Preset: %s", id(test_fan).get_preset_mode().c_str());
 
       // Migration guide: Logging option 2
       // Use %.*s format (safer, no null-termination assumption)
-      auto preset_view = id(test_fan).get_preset_mode();
-      ESP_LOGD("test", "Preset: %.*s", (int)preset_view.size(), preset_view.data());
+      auto preset_ref = id(test_fan).get_preset_mode();
+      ESP_LOGD("test", "Preset: %.*s", (int)preset_ref.size(), preset_ref.c_str());
 
       // Test != comparison
       if (id(test_fan).get_preset_mode() != "Sleep") {

